distribution/sc-proxy-dex/src/wrapped_token_merge.rs

```diff
@@ -0,0 +1,496 @@
+                &entry.attributes.farming_token_amount,
+            );
+
+            self.locked_asset_factory(locked_asset_factory_addr.clone())
+                .depositToken(
+                    locked_asset_token.clone(),
+                    entry.attributes.farming_token_nonce,
+                    locked_token_amount,
+                )
+                .execute_on_dest_context();
+        }
+
+        self.locked_asset_factory(locked_asset_factory_addr)
+            .mergeLockedAssetTokens(OptionalArg::Some(BoxedBytes::from(ACCEPT_PAY_FUNC_NAME)))
+            .execute_on_dest_context_custom_range(|_, after| (after - 1, after))
+    }
+
+    fn merge_farm_tokens(
+        &self,
+        farm_contract: &Address,
+        tokens: &[WrappedFarmToken<Self::BigUint>],
+    ) -> GenericEsdtAmountPair<Self::BigUint> {
+        if tokens.len() == 1 {
+            let token = tokens[0].clone();
+
+            return GenericEsdtAmountPair {
+                token_id: token.attributes.farm_token_id,
+                token_nonce: token.attributes.farm_token_nonce,
+                amount: token.token_amount.amount,
+            };
+        }
+
+        for entry in tokens.iter() {
+            self.farm_contract_merge_proxy(farm_contract.clone())
+                .depositFarmToken(
+                    entry.attributes.farm_token_id.clone(),
+                    entry.attributes.farm_token_nonce,
+                    entry.token_amount.amount.clone(),
+                )
+                .execute_on_dest_context();
+        }
+
+        self.farm_contract_merge_proxy(farm_contract.clone())
+            .mergeFarmTokens(OptionalArg::Some(BoxedBytes::from(ACCEPT_PAY_FUNC_NAME)))
+            .execute_on_dest_context_custom_range(|_, after| (after - 1, after))
+    }
+
+    fn merge_farming_tokens(
+        &self,
+        tokens: &[WrappedFarmToken<Self::BigUint>],
+    ) -> SCResult<GenericEsdtAmountPair<Self::BigUint>> {
+        if tokens.len() == 1 {
+            let first_token = tokens[0].clone();
+            let farming_amount = self.rule_of_three(
+                &first_token.token_amount.amount,
+                &first_token.attributes.farm_token_amount,
+                &first_token.attributes.farming_token_amount,
+            );
+
+            return Ok(GenericEsdtAmountPair {
+                token_id: first_token.attributes.farming_token_id,
+                token_nonce: first_token.attributes.farming_token_nonce,
+                amount: farming_amount,
+            });
+        }
+
+        let farming_token_id = tokens[0].clone().attributes.farming_token_id;
+        let locked_asset_token_id = self.locked_asset_token_id().get();
+
+        if farming_token_id == locked_asset_token_id {
+            Ok(self.merge_locked_asset_tokens_from_wrapped_farm(tokens))
+        } else {
+            self.merge_wrapped_lp_tokens_from_farm(tokens)
+        }
+    }
+
+    fn merge_wrapped_lp_tokens_from_farm(
```

Comment:
I think we could separate the proxy sc in 2 and the merging as well. Thus Farming and providing liquidity should be in separate contracts / token-wrapping-merging moduls.

Reply:
this would be nice but we might get into some troubles with merging since only one contract can have creation roles for a specific nft. thus when trying to merge wrapped farm tokens, the proxy farm sc would have to do execute on dest context in order to merge wrapped lp tokens and this brings some overhead. i believe separating the merge file into two different files like wrapped_lp_token_merge.rs and wrapped_farm_token_merge.rs might be a cleaner way. what do you think?